tools/online_editor/src/editor_widget.ts

```diff
@@ -0,0 +1,551 @@
+              if (model != null) {
+                monaco.editor.setModelMarkers(model, "slint", markers);
+              }
+            }
+          });
+        }
+      }, 1);
+    }
+  }
+
+  protected maybe_update_preview_automatically() {
+    if (this.auto_compile) {
+      console.log("Refreshing auto_compile timeout.");
+      if (this._keystroke_timeout_handle != null) {
+        clearTimeout(this._keystroke_timeout_handle);
+      }
+      this._keystroke_timeout_handle = setTimeout(() => {
+        this.update_preview();
+      }, 500);
+    }
+  }
+
+  private async setup_editor(container: HTMLDivElement): Promise<void> {
+    container.className += "edit-area";
+
+    monaco.languages.register({
+      id: "slint",
+      extensions: [".slint"],
+      aliases: ["Slint", "slint"],
+      mimetypes: ["application/slint"],
+    });
+    monaco.languages.onLanguage("slint", () => {
+      monaco.languages.setMonarchTokensProvider("slint", slint_language);
+    });
+
+    this._editor = monaco.editor.create(container, {
+      language: "slint",
+      glyphMargin: true,
+      lightbulb: {
+        enabled: true,
+      },
+    });
+    MonacoServices.install();
+
+    function createLanguageClient(
+      transports: MessageTransports
+    ): MonacoLanguageClient {
+      return new MonacoLanguageClient({
+        name: "Slint Language Client",
+        clientOptions: {
+          // use a language id as a document selector
+          documentSelector: [{ language: "slint" }],
+          // disable the default error handler
+          errorHandler: {
+            error: () => ({ action: ErrorAction.Continue }),
+            closed: () => ({ action: CloseAction.DoNotRestart }),
+          },
+        },
+        // create a language client connection to the server running in the web worker
+        connectionProvider: {
+          get: (_encoding: string) => {
+            return Promise.resolve(transports);
+          },
+        },
+      });
+    }
+
+    const lsp_worker = new Worker(
+      new URL("worker/lsp_worker.ts", import.meta.url),
+      { type: "module" }
+    );
+    let worker_is_set_up = false;
+    lsp_worker.onmessage = (m) => {
+      // We cannot start sending messages to the client before we start listening which
+      // the server only does in a future after the wasm is loaded.
+      if (m.data === "OK") {
+        const reader = new BrowserMessageReader(lsp_worker);
+        const writer = new BrowserMessageWriter(lsp_worker);
+
+        const languageClient = createLanguageClient({ reader, writer });
+
+        languageClient.start();
+
+        reader.onClose(() => languageClient.stop());
+
+        worker_is_set_up = true;
+      }
+    };
+
+    // Wait a bit for the worker to come up... leaving the loader in place;-)
+    for (let i = 0; i < 10; ++i) {
+      if (!worker_is_set_up) {
+        await new Promise((f) => setTimeout(f, 1000));
+      } else {
+        break;
+      }
+    }
+  }
```

Comment:
I think it would be better to use a Promise that resolves when the actual "OK" message arrives, instead of waiting for up to 10 seconds. Something like this (pseudo code):

```javascript
return new Promise((resolve) => {
    ...
    lsp_worker.onmessage = (m) => {
        if (m.data == "OK") {
            resolve();
        }
    }
});
```

And then let the caller of `setup_editor` await and remove the loader.

That said, it seems to me that while this is an async function, the call site isn't?

Reply:
That the caller is sync was exactly the problem I had:-/

Reply:
Can't you make it "async" in the same way? Imagine this function returning a promise that resolves when the worker is ready, and your caller is sync. So you could fall back to something like this:

```javascript
editor_widget.setup().then(() => { welcome_widget.hide(); /* or whatever you need to do */ }
```